plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/CodeRange.hs

```diff
@@ -0,0 +1,109 @@
+import           Control.Monad.Reader                    (runReader)
+import           Control.Monad.Trans.Maybe               (MaybeT (MaybeT),
+                                                          maybeToExceptT)
+import           Data.Coerce                             (coerce)
+import           Data.Data                               (Typeable)
+import           Data.Hashable
+import qualified Data.Map.Strict                         as Map
+import           Development.IDE
+import           Development.IDE.Core.Rules              (toIdeResult)
+import qualified Development.IDE.Core.Shake              as Shake
+import           Development.IDE.GHC.Compat              (Annotated,
+                                                          HieAST (..),
+                                                          HieASTs (getAsts),
+                                                          ParsedSource, RefMap)
+import           Development.IDE.GHC.Compat.Util
+import           Development.IDE.GHC.ExactPrint          (GetAnnotatedParsedSource (GetAnnotatedParsedSource))
+import           GHC.Generics                            (Generic)
+import           Ide.Plugin.SelectionRange.ASTPreProcess (PreProcessEnv (..),
+                                                          preProcessAST)
+import           Prelude                                 hiding (log)
+
+data Log = LogShake Shake.Log
+    | LogBadDependency BadDependencyLog
+    | LogNoAST
+
+instance Pretty Log where
+    pretty log = case log of
+        LogShake shakeLog                 -> pretty shakeLog
+        LogBadDependency badDependencyLog -> pretty badDependencyLog
+        LogNoAST                          -> "no HieAst exist for file"
+
+data BadDependencyLog = forall rule. Show rule => BadDependencyLog rule
+
+instance Pretty BadDependencyLog where
+    pretty (BadDependencyLog rule) = "can not get result from rule " <> pretty (show rule)
+
+data CodeRange = CodeRange Range [CodeRange]
+    deriving (Show, Generic)
+
+instance NFData CodeRange
+
+buildCodeRange :: HieAST a -> RefMap a -> Annotated ParsedSource -> CodeRange
+buildCodeRange ast refMap _ =
+    -- We work on 'HieAST', then convert it to 'CodeRange', so that applications such as selection range and folding
+    -- range don't need to care about 'HieAST'
+    -- TODO @sloorush actually use 'Annotated ParsedSource' to handle structures not in 'HieAST' properly (for example comments)
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    in simplify . astToCodeRange $ ast'
+
+astToCodeRange :: HieAST a -> CodeRange
+astToCodeRange (Node _ sp []) = CodeRange (realSrcSpanToRange sp) []
+astToCodeRange (Node _ sp children) = CodeRange (realSrcSpanToRange sp) (fmap astToCodeRange children)
+
+-- Remove redundant nodes in 'CodeRange' tree
+simplify :: CodeRange -> CodeRange
+simplify r@(CodeRange range1 [CodeRange range2 children])
```

Comment:
I would have assumed this function should be recursive? But it seems to only look at the very top level?

Reply:
Yep, the unit tests only catch the top-level case. I will fix it and add more tests.